pkg/extensions/sync/local.go

```diff
@@ -0,0 +1,276 @@
+	imageStore := registry.storeController.GetImageStore(repo)
+
+	// open tempImageStore
+	var tempRootDir string
+
+	if strings.HasSuffix(imageReference.StringWithinTransport(), reference) {
+		tempRootDir = strings.ReplaceAll(imageReference.StringWithinTransport(), fmt.Sprintf("%s:%s", repo, reference), "")
+	} else {
+		tempRootDir = strings.ReplaceAll(imageReference.StringWithinTransport(), fmt.Sprintf("%s:", repo), "")
+	}
+
+	metrics := monitoring.NewMetricsServer(false, registry.log)
+
+	tempImageStore := local.NewImageStore(tempRootDir, false,
+		storage.DefaultGCDelay, false, false, log.Logger{}, metrics, nil, nil)
+
+	defer os.RemoveAll(tempImageStore.RootDir())
+
+	registry.log.Info().Str("syncTempDir", path.Join(tempImageStore.RootDir(), repo)).Str("reference", reference).
+		Msg("pushing synced local image to local registry")
+
+	var lockLatency time.Time
+
+	manifestBlob, manifestDigest, mediaType, err := tempImageStore.GetImageManifest(repo, reference)
+	if err != nil {
+		registry.log.Error().Str("errorType", common.TypeOf(err)).
+			Err(err).Str("dir", path.Join(tempImageStore.RootDir(), repo)).Str("repo", repo).Str("reference", reference).
+			Msg("couldn't find synced manifest in temporary sync dir")
+
+		return err
+	}
+
+	// is image manifest
+	switch mediaType {
+	case ispec.MediaTypeImageManifest:
+		if err := registry.copyManifest(repo, manifestBlob, reference, tempImageStore); err != nil {
+			if errors.Is(err, zerr.ErrImageLintAnnotations) {
+				registry.log.Error().Str("errorType", common.TypeOf(err)).
+					Err(err).Msg("couldn't upload manifest because of missing annotations")
+
+				return nil
+			}
+
+			return err
+		}
+	case ispec.MediaTypeImageIndex:
+		// is image index
+		var indexManifest ispec.Index
+
+		if err := json.Unmarshal(manifestBlob, &indexManifest); err != nil {
+			registry.log.Error().Str("errorType", common.TypeOf(err)).
+				Err(err).Str("dir", path.Join(tempImageStore.RootDir(), repo)).
+				Msg("invalid JSON")
+
+			return err
+		}
+
+		for _, manifest := range indexManifest.Manifests {
+			tempImageStore.RLock(&lockLatency)
+			manifestBuf, err := tempImageStore.GetBlobContent(repo, manifest.Digest)
+			tempImageStore.RUnlock(&lockLatency)
+
+			if err != nil {
+				registry.log.Error().Str("errorType", common.TypeOf(err)).
+					Err(err).Str("dir", path.Join(tempImageStore.RootDir(), repo)).Str("digest", manifest.Digest.String()).
+					Msg("couldn't find manifest which is part of an image index")
+
+				return err
+			}
+
+			if err := registry.copyManifest(repo, manifestBuf, manifest.Digest.String(),
+				tempImageStore); err != nil {
+				if errors.Is(err, zerr.ErrImageLintAnnotations) {
+					registry.log.Error().Str("errorType", common.TypeOf(err)).
+						Err(err).Msg("couldn't upload manifest because of missing annotations")
+
+					return nil
+				}
+
+				return err
+			}
+		}
+
+		_, _, err = imageStore.PutImageManifest(repo, reference, mediaType, manifestBlob)
```

Comment:
No changes needed here for signatures?

Reply:
no, at the moment, cosign tags are skipped from this logic, and notation signatures don't have tags so they are also skipped from this logic.